File: jcha752.svc.cs
//Name: Jimmy Chun Yew Chan
//ID: 8302265
//UPI: JCHA752

using System;
using System.Collections.Generic;
using System.Data.Services;
using System.Data.Services.Common;
using System.Data.Services.Providers;
using System.Linq;
using System.ServiceModel;
using System.ServiceModel.Web;
using System.Web;
using System.Xml.Linq;

namespace A4CS {
    [ServiceBehavior(IncludeExceptionDetailInFaults = true)]
    public class A4CS : DataService<MyDataSource> {
        public static void InitializeService(DataServiceConfiguration config) {
            config.SetEntitySetAccessRule("*", EntitySetRights.AllRead);
            // config.SetServiceOperationAccessRule("MyServiceOperation", ServiceOperationRights.All);
            config.DataServiceBehavior.MaxProtocolVersion = DataServiceProtocolVersion.V3;
            config.UseVerboseErrors = true;
        }
    }
    //Customer Class
    [DataServiceKey("CustomerID")]
    public class Customer {
        public string CustomerID { get; set; }
        public string CompanyName { get; set; }
        public string ContactName { get; set; }
        public IEnumerable<Order> Orders { get; set; }
    }

    //Order Class
    [DataServiceKey("OrderID")]
    public class Order {
        public int OrderID { get; set; }
        public string CustomerID { get; set; }
        public DateTime? OrderDate { get; set; }
        public DateTime? ShippedDate { get; set; }
        public decimal? Freight { get; set; }
        public string ShipName { get; set; }
        public string ShipCity { get; set; }
        public string ShipCountry { get; set; }
        public Customer Customer { get; set; }
    }

    public class MyDataSource {
        static string FOLDER = @".\data\";

        static MyDataSource() {
            Console.WriteLine(" ... loading {FOLDER}\\XCustomers.xml");
            _MyCustomers = 
                XElement.Load(FOLDER + @"\XCustomers.xml")
                .Elements("Customer")
                .Select(x => new Customer {
                    CustomerID = (string)x.Element("CustomerID"),
                    CompanyName = (string)x.Element("CompanyName"),
                    ContactName = (string)x.Element("ContactName"),
                }).ToArray();

            Console.WriteLine("... loading {FOLDER}\\XOrders.xml");
            _MyOrders =
                XElement.Load(FOLDER + @"\XOrders.xml")
                .Elements("Order")
                .Select(x => new Order {
                    OrderID = (int)x.Element("OrderID"),
                    CustomerID = (string)x.Element("CustomerID"),
                    OrderDate = string.IsNullOrEmpty((string)x.Element("OrderDate")) ? (DateTime?)null : (DateTime?)x.Element("OrderDate"),
                    ShippedDate = string.IsNullOrEmpty((string)x.Element("ShippedDate")) ? (DateTime?)null : (DateTime?)x.Element("ShippedDate"),
                    Freight = string.IsNullOrEmpty((string)x.Element("Freight")) ? (decimal?)null : (decimal?)x.Element("Freight"),
                    ShipName = (string)x.Element("ShipName"),
                    ShipCity = (string)x.Element("ShipCity"),
                    ShipCountry = (string)x.Element("ShipCountry"),
                }).ToArray();

            Console.WriteLine(" ... relating _MyCustomers with _MyOrders");

            var ordersLookup = _MyOrders.ToLookup( order_cust => order_cust.CustomerID);
            var customerDictionary = _MyCustomers.ToDictionary(customer => customer.CustomerID);
            foreach (var order_cust in _MyOrders) order_cust.Customer = customerDictionary[order_cust.CustomerID];
            foreach (var customer in _MyCustomers) customer.Orders = ordersLookup[customer.CustomerID];  
            Console.WriteLine(" ... starting");
        }
        
        //Enumerables Customer
        static IEnumerable<Customer> _MyCustomers;
        //Enumerables Order
        static IEnumerable<Order> _MyOrders;
        //Queryables Customer
        public IQueryable<Customer> Customers {
            get {
                Console.WriteLine("... returning Customers");
                return _MyCustomers.AsQueryable();
            }
        }
        //Queryables Order
        public IQueryable<Order> Orders {
            get {
                Console.WriteLine("... returning Orders");
                return _MyOrders.AsQueryable();
            }
        }
    }	
}

File: jcha752.svc.fs
//Name: Jimmy Chun Yew Chan
//UPI: JCHA752
//ID: 8302265

namespace A4FS

open System
open System.Collections.Generic
open System.Data.Services
open System.Data.Services.Common
open System.Data.Services.Providers
open System.Linq
open System.ServiceModel
open System.ServiceModel.Web
open System.Web
open System.Xml.Linq

//Customer Class
type [<AllowNullLiteral(); DataServiceKey("CustomerID")>] 
  Customer () = class  // public
    member val CustomerID: string = null with get, set
    member val CompanyName: string = null with get, set
    member val ContactName: string = null with get, set
    member val Orders: seq<Order> = null with get, set
    end

//Order Class
and [<AllowNullLiteral(); DataServiceKey("OrderID")>] 
  Order () = class  // public
    member val OrderID: int = 0 with get, set
    member val CustomerID: string = null with get, set
    member val OrderDate: Nullable<DateTime> = Nullable<DateTime>() with get, set
    member val ShippedDate: Nullable<DateTime> = Nullable<DateTime>() with get, set
    member val Freight: Nullable<decimal> = Nullable<decimal>() with get, set
    member val ShipName: string = null with get, set
    member val ShipCity: string = null with get, set
    member val ShipCountry: string = null with get, set
    member val Customer: Customer = null with get, set
    end

type public MyDataSource () = class
    static let FOLDER = @".\data\"
    // @"C:\usertmp\";

    static let mutable _MyCustomers: seq<Customer> = null
    static let mutable _MyOrders: seq<Order> = null
    static let xn = XName.Get
    
    static do  // MyDataSource ()
        Console.WriteLine("... loading {0}\\XCustomers.xml", FOLDER)
        _MyCustomers <- XElement
            .Load(FOLDER + @"\XCustomers.xml")
            .Elements(xn "Customer")
            .Select(fun x -> 
                new Customer (
                    CustomerID = x.Element(xn "CustomerID").Value,
                    CompanyName = x.Element(xn "CompanyName").Value,
                    ContactName = x.Element(xn "ContactName").Value))
            .ToArray()

        Console.WriteLine("... loading {0}\\XOrders.xml", FOLDER)
        _MyOrders <- XElement
            .Load(FOLDER + @"\XOrders.xml")
            .Elements(xn "Order")
            .Select(fun x -> 
                new Order (
                    OrderID = Int32.Parse (x.Element(xn "OrderID").Value),
                    CustomerID = x.Element(xn "CustomerID").Value,
                    OrderDate = (if isNull(x.Element(xn "OrderDate")) then Nullable<DateTime>() else (if String.IsNullOrEmpty(x.Element(xn "OrderDate").Value) then Nullable<DateTime>() else Nullable (DateTime.Parse (x.Element(xn "OrderDate").Value)))),
                    ShippedDate = (if isNull(x.Element(xn "ShippedDate")) then Nullable<DateTime>() else (if String.IsNullOrEmpty(x.Element(xn "ShippedDate").Value) then Nullable<DateTime>() else Nullable (DateTime.Parse (x.Element(xn "ShippedDate").Value)))),
                    Freight = (if isNull(x.Element(xn "Freight")) then Nullable<decimal>() else (if String.IsNullOrEmpty(x.Element(xn "Freight").Value) then Nullable<decimal>() else Nullable (Decimal.Parse (x.Element(xn "Freight").Value)))),
                    ShipName = x.Element(xn "ShipName").Value,
                    ShipCity = x.Element(xn "ShipCity").Value,
                    ShipCountry = x.Element(xn "ShipCountry").Value))
            .ToArray();

        Console.WriteLine("... relating _MyCustomers with _MyOrders")

        let MyOrderLookUp = _MyOrders.ToLookup (fun order -> order.CustomerID)
        let MyCustomerDictionary = _MyCustomers.ToDictionary (fun customer -> customer.CustomerID)
        for order in _MyOrders do order.Customer <- MyCustomerDictionary.[order.CustomerID]
        for customer in _MyCustomers do customer.Orders <- MyOrderLookUp.[customer.CustomerID]
        Console.WriteLine("... starting");

    //Queryables Customers	
    member this.Customers
        with get (): IQueryable<Customer> = 
            Console.WriteLine("... returning Customers")
            _MyCustomers.AsQueryable()
    //Queryables Orders		
    member this.Orders
        with get (): IQueryable<Order> = 
            Console.WriteLine("... returning Orders")
            _MyOrders.AsQueryable()
    end

[<ServiceBehavior(IncludeExceptionDetailInFaults = true)>]
type public A4FS () = class 
    inherit DataService<MyDataSource> ()
    static member InitializeService (config: DataServiceConfiguration): unit =
        config.SetEntitySetAccessRule ("*", EntitySetRights.AllRead)
        // config.SetServiceOperationAccessRule ("MyServiceOperation", ServiceOperationRights.All)
        config.DataServiceBehavior.MaxProtocolVersion <- DataServiceProtocolVersion.V3;
        config.UseVerboseErrors <- true;
    end 